src/messaging/tests/echo/echo_requester.cpp

```diff
@@ -0,0 +1,234 @@
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    else
+    {
+        // Add some application payload data in the buffer.
+        char * p    = reinterpret_cast<char *>(payloadBuf->Start());
+        int32_t len = snprintf(p, CHIP_SYSTEM_CONFIG_HEADER_RESERVE_SIZE, "Echo Message %" PRIu64 "\n", gEchoCount);
+
+        // Set the datalength in the buffer appropriately.
+        payloadBuf->SetDataLength((uint16_t) len);
+    }
+
+    gLastEchoTime = Now();
+
+    printf("\nSend echo request message to Node: %lu\n", kServerDeviceId);
+
+    err = gEchoClient.SendEchoRequest(kServerDeviceId, std::move(payloadBuf));
+
+    if (err == CHIP_NO_ERROR)
+    {
+        gWaitingForEchoResp = true;
+        gEchoCount++;
+    }
+    else
+    {
+        printf("Send echo request failed, err: %s\n", ErrorStr(err));
+    }
+
+    return err;
+}
+
+CHIP_ERROR EstablishSecureSession()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    SecurePairingUsingTestSecret * testSecurePairingSecret = Platform::New<SecurePairingUsingTestSecret>(
+        Optional<NodeId>::Value(kServerDeviceId), static_cast<uint16_t>(0), static_cast<uint16_t>(0));
+    VerifyOrExit(testSecurePairingSecret != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    // Attempt to connect to the peer.
+    err = gSessionManager.NewPairing(
+        Optional<Transport::PeerAddress>::Value(Transport::PeerAddress::UDP(gDestAddr, CHIP_PORT, INET_NULL_INTERFACEID)),
+        kServerDeviceId, testSecurePairingSecret);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        printf("Establish secure session failed, err: %s\n", ErrorStr(err));
+        gLastEchoTime = Now();
+    }
+    else
+    {
+        printf("Establish secure session succeeded\n");
+    }
+
+    return err;
+}
+
+void HandleEchoResponseReceived(NodeId nodeId, System::PacketBufferHandle payload)
+{
+    uint32_t respTime    = Now();
+    uint32_t transitTime = respTime - gLastEchoTime;
+
+    gWaitingForEchoResp = false;
+    gEchoRespCount++;
+
+    printf("Echo Response from node %lu : %" PRIu64 "/%" PRIu64 "(%.2f%%) len=%u time=%.3fms\n", nodeId, gEchoRespCount, gEchoCount,
+           static_cast<double>(gEchoRespCount) * 100 / gEchoCount, payload->DataLength(), static_cast<double>(transitTime) / 1000);
+}
+
+} // namespace
+
+int main(int argc, char * argv[])
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (argc <= 1)
+    {
+        printf("Missing Echo Server IP address\n");
+        ExitNow(err = CHIP_ERROR_INVALID_ARGUMENT);
+    }
+
+    if (!Inet::IPAddress::FromString(argv[1], gDestAddr))
+    {
+        printf("Invalid Echo Server IP address: %s\n", argv[1]);
+        ExitNow(err = CHIP_ERROR_INVALID_ARGUMENT);
+    }
+
+    InitializeChip();
+
+    err = gTransportManager.Init(Transport::UdpListenParameters(&DeviceLayer::InetLayer)
+                                     .SetAddressType(Inet::kIPAddressType_IPv4)
+                                     .SetListenPort(ECHO_CLIENT_PORT));
```

Comment:
is this required?  can the client be on an ephemeral port?

Reply:
This is to make sure the client get set to a different port other than default CHIP_PORT, we need this if we want to test echo on the same machine.